Reflected infix append function causes "Error: Cannot parse specification"

The code below should be accepted by Liquid Haskell but is not. Removing the function `nilAppendId` or using an prefix append function instead, results in the code being accepted.

``` haskell
module Bug where
{-@ LIQUID "--higherorder"    @-}
{-@ LIQUID "--exact-data-con" @-}

import Language.Haskell.Liquid.ProofCombinators 
import Prelude hiding ((++))

{-@ infix ++ @-}
{-@ reflect ++ @-}
(++) :: [a]-> [a] -> [a] 
[] ++ ys = ys 
(x:xs) ++ ys = x : (xs ++ ys)

{-@ appendNilId :: xs:[a] -> { xs ++ [] = xs } @-}
appendNilId :: [a] -> Proof
appendNilId xs = undefined 

{-@ nilAppendId :: xs:[a] -> { [] ++ xs = xs } @-}
nilAppendId :: [a] -> Proof
nilAppendId xs = undefined 
```

### Error message

```
LiquidHaskell Version 0.8.4.0, Git revision ea8217bcdd78e550ef2b39237dc83fe2f1122b69
Copyright 2013-18 Regents of the University of California. All Rights Reserved.

**** DONE:  annotate ***********************************************************
 
**** RESULT: ERROR *************************************************************

 [PATH TO FILE]/Bug.hs:19:35: Error: Cannot parse specification:
  
 19 | {-@ nilAppendId :: xs:[a] -> { [] ++ xs = xs } @-}
                                        ^
     unexpected "+"
     expecting white space or "}"
```

--- a/src/Language/Haskell/Liquid/Parse.hs
+++ b/src/Language/Haskell/Liquid/Parse.hs
@@ -324,7 +324,7 @@ bareTypeBracesP = do
                return $ Right ct
                      ))
            <|>
-            (try(do
+            (do
                     x  <- symbolP
                     _ <- colon
                     -- NOSUBST i  <- freshIntP
@@ -335,17 +335,15 @@ bareTypeBracesP = do
                     -- su replaces any use of x in the balance of the expression with the unique val
                     -- NOSUBST let xi = intSymbol x i
                     -- NOSUBST let su v = if v == x then xi else v
-                    return $ Left $ PC (PcExplicit x) $ t (Reft (x, ra)) ))
-           <|> do t <- ((RHole . uTop . Reft . ("VV",)) <$> (refasHoleP <* spaces))
-                  return (Left $ nullPC t)
-            )) <|> try (helper holeOrPreds) <|> helper predP
+                    return $ Left $ PC (PcExplicit x) $ t (Reft (x, ra)) )
+            )) <|> try (helper holeOrPredsP) <|> helper predP
   case t of
     Left l -> return l
     Right ct -> do
       PC _sb tt <- btP
       return $ nullPC $ rrTy ct tt
   where
-    holeOrPreds
+    holeOrPredsP
       = (reserved "_" >> return hole)
      <|> try (pAnd <$> brackets (sepBy predP semi))
     helper p = braces $ do

--- a/src/Language/Haskell/Liquid/Parse.hs
+++ b/src/Language/Haskell/Liquid/Parse.hs
@@ -318,7 +318,7 @@ bareTypeP = do
 
 bareTypeBracesP :: Parser ParamComp
 bareTypeBracesP = do
-  t <-  braces (
+  t <-  try (braces (
             (try (do
                ct <- constraintP
                return $ Right ct
@@ -338,12 +338,20 @@ bareTypeBracesP = do
                     return $ Left $ PC (PcExplicit x) $ t (Reft (x, ra)) ))
            <|> do t <- ((RHole . uTop . Reft . ("VV",)) <$> (refasHoleP <* spaces))
                   return (Left $ nullPC t)
-            )
+            )) <|> try (helper holeOrPreds) <|> helper predP
   case t of
     Left l -> return l
     Right ct -> do
       PC _sb tt <- btP
       return $ nullPC $ rrTy ct tt
+  where
+    holeOrPreds
+      = (reserved "_" >> return hole)
+     <|> try (pAnd <$> brackets (sepBy predP semi))
+    helper p = braces $ do
+      t <- ((RHole . uTop . Reft . ("VV",)) <$> (p <* spaces))
+      return (Left $ nullPC t)
+
 
 bareArgP :: Symbol -> Parser BareType
 bareArgP vvv

--- a/tests/parser/pos/ReflectedInfix.hs
+++ b/tests/parser/pos/ReflectedInfix.hs
@@ -1,0 +1,20 @@
+module ReflectedInfix where
+{-@ LIQUID "--higherorder"    @-}
+{-@ LIQUID "--exact-data-con" @-}
+
+import Language.Haskell.Liquid.ProofCombinators 
+import Prelude hiding ((++))
+
+{-@ infix ++ @-}
+{-@ reflect ++ @-}
+(++) :: [a]-> [a] -> [a] 
+[] ++ ys = ys 
+(x:xs) ++ ys = x : (xs ++ ys)
+
+{-@ appendNilId :: xs:[a] -> { xs ++ [] = xs } @-}
+appendNilId :: [a] -> Proof
+appendNilId xs = undefined 
+
+{-@ nilAppendId :: xs:[a] -> { [] ++ xs = xs } @-}
+nilAppendId :: [a] -> Proof
+nilAppendId xs = undefined 

